core/watcher/watchertest/secrets.go

```diff
@@ -5,128 +5,49 @@ package watchertest
+	return func(c *gc.C, received [][]watcher.SecretTriggerChange) bool {
+		flattened := utils.Flatten(received)
+		if len(flattened) >= len(expected) {
+			mc := jc.NewMultiChecker()
+			mc.AddExpr(`_[_].NextTriggerTime`, jc.Almost, jc.ExpectedValue)
+			c.Assert(flattened, mc, expected)
+			return true
 		}
-	}
-}
-
-func (c SecretsTriggerWatcherC) AssertClosed() {
-	select {
-	case _, ok := <-c.Watcher.Changes():
-		c.Assert(ok, jc.IsFalse)
-	default:
-		c.Fatalf("watcher not closed")
-	}
-}
-
-// SecretBackendRotateWatcherC embeds a gocheck.C and adds methods to help
-// verify the behaviour of any watcher that uses a
-// <-chan []SecretBackendRotateChange
-type SecretBackendRotateWatcherC struct {
-	*gc.C
-	Watcher watcher.SecretBackendRotateWatcher
-}
-
-// NewSecretBackendRotateWatcherC returns a SecretBackendRotateWatcherC that
-// checks for aggressive event coalescence.
-func NewSecretBackendRotateWatcherC(c *gc.C, w watcher.SecretBackendRotateWatcher) SecretBackendRotateWatcherC {
-	return SecretBackendRotateWatcherC{
-		C:       c,
-		Watcher: w,
-	}
-}
-
-func (c SecretBackendRotateWatcherC) AssertNoChange() {
-	select {
-	case actual, ok := <-c.Watcher.Changes():
-		c.Fatalf("watcher sent unexpected change: (%v, %v)", actual, ok)
-	case <-time.After(testing.ShortWait):
-	}
-}
-
-// AssertChanges asserts the given changes was reported by the watcher,
-// but does not assume there are no following changes.
-func (c SecretBackendRotateWatcherC) AssertChanges(expect ...watcher.SecretBackendRotateChange) {
-	var received []watcher.SecretBackendRotateChange
-	timeout := time.After(testing.LongWait)
-	for a := testing.LongAttempt.Start(); a.Next(); {
-		select {
-		case actual, ok := <-c.Watcher.Changes():
-			c.Logf("Secrets Trigger Watcher.Changes() => %# v", actual)
-			c.Assert(ok, jc.IsTrue)
-			sort.Slice(actual, func(i, j int) bool {
-				return actual[i].Name < actual[j].Name
-			})
-			received = append(received, actual...)
-			if len(received) >= len(expect) {
-				mc := jc.NewMultiChecker()
-				mc.AddExpr(`_[_].NextTriggerTime`, jc.Almost, jc.ExpectedValue)
-				c.Assert(received, mc, expect)
-				return
-			}
-		case <-timeout:
-			c.Fatalf("watcher did not send change")
+		return false
+	}
+}
+
+// SecretBackendRotateTriggerTimeAssert is an assert function used as parameter
+// for the generic WatcherC. It asserts that the `NextTriggerTime` in the
+// `SecretBackendRotateChange` is "almost" as the passed expected value.
+// It returns true if the assertion is correct (meaning that the listening
+// loop in WatcherC can break).
+func SecretBackendRotateTriggerTimeAssert(expected ...watcher.SecretBackendRotateChange) func(c *gc.C, received [][]watcher.SecretBackendRotateChange) bool {
+	return func(c *gc.C, received [][]watcher.SecretBackendRotateChange) bool {
+		flattened := utils.Flatten(received)
+		sort.Slice(flattened, func(i, j int) bool {
```

Comment:
**issue(maybe):** `sort.Slice` may not be stable. Since we are comparing times, which I assumes are ordered, this instruction may cause random failure of test.

**suggestion:** use `sort.SliceStable`